pip/pip-379.md

```diff
@@ -0,0 +1,115 @@
+
+### High-Level implementation plan
+
+The proposed solution introduces a "draining hashes" concept to efficiently manage message ordering in Key_Shared subscriptions:
+
+1. When consumer hash ranges change (e.g., a consumer joins or leaves), affected hashes of pending messages are added to a "draining hashes" set.
+2. Following messages with hashes in the "draining hashes" set are blocked from further delivery until pending messages are processed.
+3. A reference counter tracks pending messages for each hash in the "draining hashes" set.
+4. As messages are acknowledged or consumers disconnect and therefore get removed from pending messages, the reference counter is decremented.
+5. When the reference counter reaches zero, the hash is removed from the set, allowing new message delivery.
+6. Consumer hash assignments may change multiple times, and a draining hash might be reassigned to the original consumer. The draining hash data structure contains information about the draining consumer. When a message is attempted for delivery, the system can check if the target consumer is the same as the draining consumer. If they match, there's no need to block the hash. The implementation should also remove such hashes from the draining hashes set. This "lazy" approach reduces the need for actively scanning all draining hashes whenever hash assignments change.
+
+This approach will meet the updated contract of preserving ordering while minimizing the impact on performance and memory usage. The tracking only comes into play during transition states. When consumers have been connected for a longer duration and all draining hashes have been removed, there won't be a need to check any special rules or maintain any extra state.
+
+## Public-facing Changes
+
+### Topic Stats Changes & Observability
+
+Topic stats for the removed PIP-282 "recently joined consumers"/"last sent position" solution are removed:
+- `lastSentPositionWhenJoining` field for each consumer
+- `consumersAfterMarkDeletePosition` field for each Key_Shared subscription
```

Comment:
This is a break change for the existing users for how they troubleshoot Key_Shared subscription issues today.

Reply:
There's a replacement. It's not a breaking change in that sense. That's also the reason why there's a benefit that this goes into 4.0 since in major versions breaking changes are expected. I'm sure that nobody will miss using these fields! If they do, they should remain on Pulsar 3.x.

There won't even be a need to troubleshoot consumer stuck issues in PIP-379 design in the same way. There's a replacement for cases where a consumer really doesn't acknowledge a message. There's also an optional REST API for finding out the actual message IDs of the unacked messages that could be used to assist the troubleshooting.

Reply:
That is a break change, we changed the REST API without any smooth transition. The field `consumersAfterMarkDeletePosition` will be delete, some users might build metrics or something on this field. We need to keep the default behavior consistent for all the APIs exposed to users. We can add more fields, but should not remove the existing one or change the meaning of the existing one.

3.0 to 4.0 should not have break change, we promised user that they upgrade from 3.0 to 4.0 directly. https://pulsar.apache.org/contribute/release-policy/#compatibility-between-releases.

Reply:
> That is a break change, we changed the REST API without any smooth transition. The field `consumersAfterMarkDeletePosition` will be delete, some users might build metrics or something on this field. We need to keep the default behavior consistent for all the APIs exposed to users. We can add more fields, but should not remove the existing one or change the meaning of the existing one.
> 
> 3.0 to 4.0 should not have break change, we promised user that they can upgrade from 3.0 to 4.0 directly without any compatibility issue. https://pulsar.apache.org/contribute/release-policy/#compatibility-between-releases. 

It's change, but not a breaking change. This is not a compatibility in the sense that our release policy expresses. That's about bookkeeper storage formats and such major details.

Reply:
I can't agree more. If we follow this way, we can delete or modify any output for the REST API.

Reply:
@codelipenghui [PIP-282](https://github.com/apache/pulsar/blob/master/pip/pip-282.md) already changed the fields and the meaning of the fields. Have you considered that?

Reply:
Yes, the proposal [PIP-282](https://github.com/apache/pulsar/blob/master/pip/pip-282.md) should provide an option for user to move smoothly. Instead of changing meaning or rename the output of public API directly.

Reply:
@codelipenghui  I've added a comment this in the update to the mailing list, https://lists.apache.org/thread/vrsjzkk3lnzq2dqhg380740hr9ftdn4t.